src/OpenTelemetry.Extensions.Hosting/OpenTelemetryBuilder.cs

```diff
@@ -95,7 +118,26 @@ public OpenTelemetryBuilder WithMetrics(Action<MeterProviderBuilder> configure)
 
         var builder = new MeterProviderBuilderBase(this.Services);
 
-        configure(builder);
+        builder.Configure(configure);
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds metric services into the builder.
+    /// </summary>
+    /// <remarks><inheritdoc cref="WithMetrics()" path="/remarks"/></remarks>
+    /// <param name="configure"><see cref="MeterProviderBuilder"/>
+    /// configuration callback.</param>
+    /// <returns>The supplied <see cref="OpenTelemetryBuilder"/> for chaining
+    /// calls.</returns>
+    public OpenTelemetryBuilder WithMetrics(Action<MeterProviderBuilder, IServiceProvider> configure)
```

Comment:
I think we need to remove this. It just won't work for most everything that exists 😢 For example, try this...

`services.AddOpenTelemetry().WithMetrics((builder, sp) => builder.AddOtlpExporter())`

You should get a `NotSupportedException` thrown. The reason is the extensions need the service collection to do options work and it can't be done once the `IServiceProvider` has been created.

There are other existing APIs you can use to accomplish this where it is needed. IMO having this API as a top-level thing will just lead people into trouble.

Reply:
> There are other existing APIs you can use to accomplish this where it is needed. IMO having this API as a top-level thing will just lead people into trouble.

Sadly not, as to get to services form the DI, you know, you need the DI. So basically I probably have to reimplement the whole Hosting stuff... Yay. Looks like the current architecture sadly doesn't seem to be suitable to that... Our specific use cases needs stuff from the DI to feed into the `AddOtlpExporter` so that seems to be off the table...

Might need to work out something using factories or configure methods and stuff like that, but that'll probably be more complex and would need a full reengineer of the hosting and be therefore unsuitable to this project...

Reply:
Also from the looks of it it seems like it should be fine, but sadly your solution doesn't even seem to build currently:
```
CSC: Error CS1705 : Assembly 'OpenTelemetry.Instrumentation.Runtime' with identity 'OpenTelemetry.Instrumentation.Runtime, Version=1.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c' uses 'OpenTelemetry.Api, Version=1.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c' which has a higher version than referenced assembly 'OpenTelemetry.Api' with identity 'OpenTelemetry.Api, Version=0.0.0.0, Culture=neutral, PublicKeyToken=7bd6737fe5b67e3c'
```

Reply:
So after unloading some projects I got it to build and also found the cuprit, being the need to `ConfigureServices`.

So would it maybe be feasible to add the `IServiceProvider` to the `AddOtlpExporter` configuration. It's kinda not as consistent, but might be actually feasible, as there's a service provider available `AddReader`.

Reply:
I'm totally open to making changes! Can you show me what you are trying to do with service provider + AddOtlpExporter? Just curious what the use case is because I haven't seen anyone ask for this yet (that I can recall).

Reply:
Well basically because we have a base class throughout a bunch of services, we need to keep stuff pretty generic. For the resources (which should be working already with this PR as you noted) the code would e similar to the following:
```csharp
.ConfigureResource((builder, sp) => {
    var svc = sp.GetRequiredService<Service>();
    builder.AddService("...", instanceId: svc.GetUniqueIdKnowToUs());
})
```

Pretty similar for the tracing:
```csharp
.WithTracing((builder, sp) => {
    var svc = sp.GetRequiredService<Service>();
    builder.AddStuffInstrumentation()
        .AddOtelExporter(options => {
            options.Uri = svc.FetchUriForOtel();
        });
})
```

Reply:
I do have to test, whether this might work:
```csharp
services.AddOptions<OtlpExporterOptions>().Configure<Service>((opts, svc) => {
    // ...
});
```

Reply:
It will, yes! If you just want to configure options, we fully support the Options API (as of 1.4). Just watch out because OtlpExporter has the same options class shared for traces and metrics. If you want to do this style and do individual things for traces & metrics use named options a la:

services.AddOptions<OtlpExporterOptions>().Configure<Service>("traces", (opts, svc) => {... });
services.AddOptions<OtlpExporterOptions>().Configure<Service>("metrics", (opts, svc) => {... });

tracerBuilder.AddOtlpExporter(name: "traces"...);
meterBuilder.AddOtlpExporter(name: "metrics"...);

Reply:
PS: For you resource case check out #4139.

Reply:
Add docs for the exporter case, for the resource case the detector looks nice, but it doesn't seem to be documented at all and I have yet to find a way to like easily register it with Open Telemetry. It doesn't seem like it just picks them up in from the DI or so.